### PR TITLE
Fix #491

### LINKED ISSUE #491: Notice: Undefined variable: actionName
On dev-master the logic for a custom controller is not correct yet.

When defining a custom controller in the resource like this:

```php
/**
 *@Resource(itemOperations={"GET": {"controller": "test", "method": "test"}})
 */
```

You will get the following error:
*Notice: Undefined variable: actionName*

ref: src/Bridge/Symfony/Routing/ApiLoader.php:114
```php
        if (isset($operation['controller'])) {
            $controller = $operation['controller'];
        } else {
            $actionName = sprintf('%s_%s', strtolower($operation['method']), $collection ? 'collection' : 'item');
            $controller = self::DEFAULT_ACTION_PATTERN.$actionName;
        }
        $path = '/'.$normalizedShortName;
        if (!$collection) {
            $path .= '/{id}';
        }
        $routeName = sprintf('%s%s_%s', self::ROUTE_NAME_PREFIX, $normalizedShortName, $actionName);
```

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #491
| License       | MIT
| Doc PR        | ∅

```
Notice: Undefined variable: actionName
```

see https://github.com/api-platform/core/issues/491#issuecomment-206396417

### CODE REVIEW

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 1]
```diff
@@ -99,11 +99,12 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
...
 
+        $actionName = sprintf('%s_%s', strtolower($operation['method']), $collection ? 'collection' : 'item');
+        $controller = self::DEFAULT_ACTION_PATTERN.$actionName;
+
         if (isset($operation['controller'])) {
             $controller = $operation['controller'];
-        } else {
-            $actionName = sprintf('%s_%s', strtolower($operation['method']), $collection ? 'collection' : 'item');
-            $controller = self::DEFAULT_ACTION_PATTERN.$actionName;
+            $actionName .= $operation['controller'];
```
Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Not sure about this resulting in the best name, before this was defined through the `route_name` parameter.

@see: https://github.com/api-platform/core/blob/1.x/Api/Operation/OperationFactory.php#L156

**Comment by soyuka** [comment]
ping @api-platform/core-team

This should be discussed:

- is the `actionName` based on the controller safe enough? 
- what about routes name collision?
- the route `path` has to be configurable, is there anything that should be added to avoid collisions?

Rel #296

**Comment by soyuka** [comment]
Closing in favor of #490.

**Comment by teohhanhui** [comment]
@soyuka This is not related to #490

**Comment by soyuka** [comment]
It is to me, I can't add behat tests if it keep throwing the Notice when a controller is specified. I did a [better implementation](https://github.com/soyuka/DunglasApiBundle/blob/fileloader/src/Bridge/Symfony/Routing/ApiLoader.php) there, this is why I closed this one.

**Comment by teohhanhui** [comment]
@soyuka I understand what you're facing, and I know it can be frustrating. But this is a bug fix. It should not be part of #490

**Comment by teohhanhui** [comment]
Do you mind porting your fix from #490 to this one? Then we can get this merged in first...

**Comment by soyuka** [comment]
There you go :)

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 2]
```diff
@@ -99,14 +99,20 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
...
 
+        $actionName = sprintf('%s_%s', strtolower($operation['method']), $collection ? 'collection' : 'item');
+        $controller = self::DEFAULT_ACTION_PATTERN.$actionName;
+
         if (isset($operation['controller'])) {
             $controller = $operation['controller'];
-        } else {
-            $actionName = sprintf('%s_%s', strtolower($operation['method']), $collection ? 'collection' : 'item');
-            $controller = self::DEFAULT_ACTION_PATTERN.$actionName;
+            $actionName .= sprintf('_%s', $operation['name'] ?? $controller);
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Isn't `$operation['name']` required, and therefore always available?

Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: I was asking myself the same thing, but I didn't find a certainty that an Operation has a name when defined through Annotations.

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 3]
```diff
@@ -99,14 +99,20 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
...
         }
 
         $path = '/'.$normalizedShortName;
+
+        if (isset($operation['path'])) {
+            $path = $operation['path'];
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: This should come after the next block.

File: src/Bridge/Symfony/Routing/ApiLoader.php (hunk 2 above)
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: It must have a name :)

Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Oh wait... The name is in `$operationName`.

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 4]
```diff
@@ -99,16 +99,22 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
...
-            $controller = self::DEFAULT_ACTION_PATTERN.$actionName;
+            $actionName .= sprintf('_%s', $operation['name']);
         }
 
-        $path = '/'.$normalizedShortName;
-        if (!$collection) {
-            $path .= '/{id}';
+        $path = $operation['path'] ?? null;
+
+        if(null === $path) {
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: CS

File: src/Bridge/Symfony/Routing/ApiLoader.php (hunk 2 above)
Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: If it's the array key, it'll be the method Name... https://github.com/api-platform/core/blob/master/src/Metadata/Resource/Factory/OperationResourceMetadataFactory.php#L58

Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Yes, that is the case for the default operations. For custom operations, see https://github.com/api-platform/core/issues/491#issuecomment-206274755

Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: This is also why it's kinda related with #490, I made sure that every declared Operation will be named (either the Yaml operation `key`, either the Xml `name` attribute)

Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Sorry, I don't see how it's related. For operations declared via annotations, there has always been a name, just like the defaults.

Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Yes my bad, I've fixed it in the last commit ;) (and I'm pushing the #490 to fix this)

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 5]
```diff
@@ -99,16 +99,22 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
             throw new RuntimeException('Either a "route_name" or a "method" operation attribute must exist.');
         }
 
+        $actionName = sprintf('%s_%s', strtolower($operation['method']), $collection ? 'collection' : 'item');
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: We can simplify it like this:

```php
$actionName = sprintf('%s_%s', strtolower($operationName), $collection ? 'collection' : 'item');
$controller = $operation['controller'] ?? self::DEFAULT_ACTION_PATTERN.$actionName;
```

This means if you set a custom operation name that's different from the method, it'll not match the default action. But I think that's not unreasonable.

**Comment by teohhanhui** [comment]
I'm :+1: once the commented items are fixed.

ping @api-platform/core-team

**Comment by teohhanhui** [comment]
Oh yes, do you think you can add some tests too? :)

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 6]
```diff
@@ -99,16 +99,19 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
...
-        if (isset($operation['controller'])) {
-            $controller = $operation['controller'];
-        } else {
-            $actionName = sprintf('%s_%s', strtolower($operation['method']), $collection ? 'collection' : 'item');
-            $controller = self::DEFAULT_ACTION_PATTERN.$actionName;
-        }
+        $collectionString = $collection ? 'collection' : 'item';
+
+        $actionName = sprintf('%s_%s', $operationName, $collectionString);
+        $controller = $operation['controller'] ?? self::DEFAULT_ACTION_PATTERN.sprintf('%s_%s', strtolower($operation['method']), $collectionString);
```
Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: I had to do this or without controller it won't keep the default controller but add a non-existant one.

Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: What do you mean?

Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: As long as `$operationName` matches `strtolower($operation['method'])`, it should be fine.

Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: With my code:

```php
$actionName = sprintf('%s_%s', strtolower($operationName), $collection ? 'collection' : 'item');
$controller = $operation['controller'] ?? self::DEFAULT_ACTION_PATTERN.$actionName;
```

This is supported:

```php
/**
 * @Resource(itemOperations={
 *     "get"={"method"="GET"},
 * })
 */
```

while this is not:

```php
/**
 * @Resource(itemOperations={
 *     "my_get_op"={"method"="GET"},
 * })
 */
```

but you could specify it explicitly:

```php
/**
 * @Resource(itemOperations={
 *     "my_get_op"={"method"="GET", "controller"="api_platform.action.get_item"},
 * })
 */
```

Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Indeed, but if no controller is specified it's better to keep the default
one, less error prone, don't you agree?
Also, this is the same behavior as in 1.x
Le ven. 8 avr. 2016 à 13:14, Teoh Han Hui <notifications@github.com> a
écrit :

File: tests/Bridge/Symfony/Routing/ApiLoaderTest.php [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        $resourceMetadata = $resourceMetadata->withItemOperations([
+            'get' => ['method' => 'GET'],
+            'put' => ['method' => 'PUT'],
+            'delete' => ['method' => 'DELETE'],
+        ]);
+
+        //custom operations
+        $resourceMetadata = $resourceMetadata->withCollectionOperations([
+            'my_op' => ['method' => 'GET', 'controller' => 'some.service.name'], //with controller
+            'my_second_op' => ['method' => 'POST'], //without controller, takes the default one
```
Comment (soyuka) [inline tests/Bridge/Symfony/Routing/ApiLoaderTest.php]: @teohhanhui this test for example. Here the controller will default to `api_platform.action.post_item`. IMO it's an expected behavior and this is the same behavior as in 1.x.

Comment (teohhanhui) [inline tests/Bridge/Symfony/Routing/ApiLoaderTest.php]: Okay :)

But let's try to make the code more elegant.

**Comment by soyuka** [comment]
Improved tests and code, should be better now :).

@api-platform/core-team

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 8]
```diff
@@ -99,16 +99,26 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
...
+
+        if (null === $controller) {
+            $actionName = sprintf('%s_%s', strtolower($operation['method']), $collectionString);
             $controller = self::DEFAULT_ACTION_PATTERN.$actionName;
         }
 
-        $path = '/'.$normalizedShortName;
-        if (!$collection) {
-            $path .= '/{id}';
+        if ($operationName !== strtolower($operation['method'])) {
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: `$actionName` is undefined when `$controller` is not `null` and `$operationName` matches `strtolower($operation['method'])`.

And I am not sure this code is better. It is hard to follow...

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 9]
```diff
@@ -99,16 +99,26 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
             throw new RuntimeException('Either a "route_name" or a "method" operation attribute must exist.');
         }
 
-        if (isset($operation['controller'])) {
-            $controller = $operation['controller'];
-        } else {
-            $actionName = sprintf('%s_%s', strtolower($operation['method']), $collection ? 'collection' : 'item');
+        $collectionString = $collection ? 'collection' : 'item';
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: I'd rather this be repeated. It's confusing...

Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Ok I'll try a more repeat-myself alternative

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 10]
```diff
@@ -99,16 +99,26 @@ private function addRoute(RouteCollection $routeCollection, string $resourceClas
...
         }
 
-        $path = '/'.$normalizedShortName;
-        if (!$collection) {
-            $path .= '/{id}';
+        if ($operationName !== strtolower($operation['method'])) {
+            $actionName = sprintf('%s_%s', $operationName, $collectionString);
+        }
+
+        $path = $operation['path'] ?? null;
```
Comment (theofidry) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: to make it more readable L114-L122 could be extracted in a method `getPath(array $operation): string` or `determinePath`

Comment (theofidry) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: on the same token L102-108 could be extracted in a sub-method as well to be more declaration and less imperative

File: tests/Bridge/Symfony/Routing/ApiLoaderTest.php [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+            ],
+            [],
+            [],
+            '',
+            [],
+            $methods
+        );
+    }
+
+    public function testApiLoader()
```
Comment (theofidry) [inline tests/Bridge/Symfony/Routing/ApiLoaderTest.php]: public methods should be placed before private ones

File: src/Bridge/Symfony/Routing/ApiLoader.php (hunk 8 above)
Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: there, not sure I can do better, this condition has to stay.

File: tests/Bridge/Symfony/Routing/ApiLoaderTest.php (hunk 11 above)
Comment (soyuka) [inline tests/Bridge/Symfony/Routing/ApiLoaderTest.php]: done, where can I find those CS rules?

File: src/Bridge/Symfony/Routing/ApiLoader.php (hunk 8 above)
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]:
```php
/**
 * @Resource(itemOperations={
 *     "my_get_op"={"method"="GET"},
 * })
 */
```

will produce: `api_platform.action.my_get_op_item` which is not what we want.

After giving this more thought, I think we should be asking ourselves what is the intended / expected use case for custom operations... That should inform our naming scheme. Actually, there is no harm in using the old way. I don't think there should be more than one custom operations which use the same default action.

Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: As you don't have any controller in this annotation, it'll get the `api_platform.action.get_item` controller with a route name of `api_platform.action.my_get_op_item` which is what I would expect.

Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: My bad... :+1:

**Comment by dunglas** [comment]
Thank you @soyuka
